### PR TITLE
Fix spectrum number saving loading to/from NeXus

### LINKED ISSUE #24702: Messed spectrum numbers after loading a Mantid NeXus
**Original reporter:** [nobody].

### Expected behavior

A loaded Mantid NeXus file has the same spectrum numbering as the original workspace.

### Actual behavior

In some cases the loaded workspace's spectrum numbers start from 1 (a default?) and are inequal to the original workspace's.

### Steps to reproduce the behavior

```python
fresh = CreateSampleWorkspace()
in_theta = ConvertSpectrumAxis(fresh, Target='Theta')
transposed = Transpose(in_theta)
cropped = CropWorkspace(transposed, StartWorkspaceIndex=6, EndWorkspaceIndex=6)
SaveNexus(cropped, '/users/soininen/delete_me.nxs')
loaded = Load('/users/soininen/delete_me.nxs')

print(loaded.getSpectrum(0).getSpectrumNo())
print(cropped.getSpectrum(0).getSpectrumNo())
assert(loaded.getSpectrum(0).getSpectrumNo() == cropped.getSpectrum(0).getSpectrumNo())
```

### Platforms affected

All.

### PR DESCRIPTION
**Pull request by soininen**
`SaveNexusProcessed` failed to write the spectrum number information to the output file if the input workspace did not have detectors attached to the histograms. Conversely, `LoadNexusProcessed` failed to load the spectrum numbers from such a file. This PR fixes.

**Report to:** [nobody].

**To test:**

The script from the issue description should pass after these changes are applied:

```python
fresh = CreateSampleWorkspace()
in_theta = ConvertSpectrumAxis(fresh, Target='Theta')
transposed = Transpose(in_theta)
cropped = CropWorkspace(transposed, StartWorkspaceIndex=6, EndWorkspaceIndex=6)
SaveNexus(cropped, '/users/soininen/delete_me.nxs')
loaded = Load('/users/soininen/delete_me.nxs')

print(loaded.getSpectrum(0).getSpectrumNo())
print(cropped.getSpectrum(0).getSpectrumNo())
assert(loaded.getSpectrum(0).getSpectrumNo() == cropped.getSpectrum(0).getSpectrumNo())
```

Fixes #24702.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/SaveNexusProcessed.cpp [hunk 1]
```diff
@@ -262,13 +261,18 @@ void SaveNexusProcessed::doExec(
...
     }
 
     cppFile.openGroup("instrument", "NXinstrument");
-    saveSpectraMapNexus(*matrixWorkspace, &cppFile, spec, ::NeXus::LZW);
+    cppFile.makeGroup("detector", "NXdetector", true);
+    cppFile.putAttr("version", 1);
+    saveSpectraDetectorMapNexus(*matrixWorkspace, &cppFile, indices,
+                                ::NeXus::LZW);
+    saveSpectrumNumbersNexus(*matrixWorkspace, &cppFile, indices, ::NeXus::LZW);
+    cppFile.closeGroup();
```
Comment (Mantid-Matthew) [inline Framework/DataHandling/src/SaveNexusProcessed.cpp]: Do we need to call this function twice?

File: Framework/DataHandling/src/SaveNexusProcessed.cpp [hunk 2]
```diff
@@ -639,18 +623,37 @@ void SaveNexusProcessed::saveSpectraMapNexus(
...
       }
     } else
-      for (int i = 0; i < 3 * ndet; i++)
+      for (size_t i = 0; i < 3 * nDetectors; i++)
         detPos[i] = 0.;
 
-    dims[0] = ndet;
+    dims.front() = static_cast<int>(nDetectors);
     dims.push_back(3);
-    dims[1] = 3;
```
Comment (Mantid-Matthew) [inline Framework/DataHandling/src/SaveNexusProcessed.cpp]: Why does dims[1] no longer need to be set?

**Review by Matthew-Andrew — COMMENTED** [review]
The code changes look good to me as do the added tests. The given test script passes for me locally and files which do have detectors attached still work.

I just have a few minor questions below.

File: Framework/DataHandling/src/SaveNexusProcessed.cpp (hunk 2 above)
Comment (soininen) [reply Framework/DataHandling/src/SaveNexusProcessed.cpp]: It is set by the `push_back(3)` call above.

File: Framework/DataHandling/src/SaveNexusProcessed.cpp (hunk 1 above)
Comment (soininen) [reply Framework/DataHandling/src/SaveNexusProcessed.cpp]: Yes, on line 269 we `openGroup` and immediately `makeGroup`. Both groups have to be closed before returning from this function.

**Review by Matthew-Andrew — APPROVED** [review]
Thanks for the explanations, now happy to approve :+1:

**Pull request merged**